spatialos-sdk/src/worker/schema.rs

```diff
@@ -41,3 +109,151 @@ pub trait SchemaPrimitiveField {
         result
     }
 }
+
+/// A struct that can be serialized into (and deserialized from) a [`SchemaObject`].
+///
+/// Schemalang [components] and [types], as well as the various associated types
+/// generated for components (i.e. updates and commands), are represented as a
+/// [`SchemaObject`] when serialized. This trait defines the conversion for such
+/// types, and provides a blanket implementation of the [`Field`] trait so that the
+/// types can be serialized as part of larger objects.
+///
+/// You should generally not have to implement this trait manually for any types.
+/// The implementation for any schema-defined types will be generated for you by the
+/// code generator provided with the SDK.
+///
+/// [components]: https://docs.improbable.io/reference/14.2/shared/schema/reference#components
+/// [types]: https://docs.improbable.io/reference/14.2/shared/schema/reference#types
+/// [`SchemaObject`]: struct.SchemaObject.html
+/// [`Field`]: trait.Field.html
+pub trait ObjectField: Sized + Clone {
+    /// Deserializes an instance of this type from a [`SchemaObject`].
+    ///
+    /// Note that deserialization can never fail as the underlying SpatialOS API ensures
+    /// that missing or invalid fields will always provide a valid default value during
+    /// deserialization. As such, if the data in `object` doesn't match what is expected
+    /// by the deserialization logic (e.g. due to a version mismatch between the current
+    /// schema version and the version `object` was generated from), the returned object
+    /// will have default values for any fields that are missing in `object` or have a
```

Comment:
Rather than default values, it could be garbage data I think

Reply:
Huh, is that documented anywhere? I swear I remember @dgavedissian saying you'd get default values if you asked the schema API for fields that don't exist (or asking for an existing field with the wrong type), but I could be putting words in his mouth 😛

Reply:
> for an existing field with the wrong type

This I'm pretty sure returns garbage data (I think I've ran into this before). But I'd need to double check.

For a non-existent field, I'm really not sure!

Reply:
Huh. Is there any way to verify that a given field matches the expected type? That could be another potential case for failure on deserialization 🤔

Reply:
Huh. Is there any way to verify that a given field matches the expected type? That could be another potential case for failure on deserialization 🤔

Reply:
Even worse we won't know that the deserialization has failed if we have the wrong types!

Reply:
Bear in mind that `Schema_GetXX` functions under the hood performs lazy deserialization of serialized protobuf data. Therefore, if you get the type wrong, there's a chance that it'll interpret the bytes in the wrong way (so you get garbage out). However, if you try and read a non-existent field, then it'll just return the default value (always 0 / null).

Reply:
We specify in our docs [here](https://docs.improbable.io/reference/14.2/shared/schema/evolution#making-changes-to-schema) that schema is only forwards/backwards compatible if you never reuse field IDs. So, you shouldn't expect it to really happen that component data doesn't match the schema (and the field IDs are different), unless the user has really messed up.

Reply:
What's the "official" way to handle non-compatible schema changes? Based on the lack of error handling in the C API and the things you've said, it seems like it comes down to:

* Detectable errors (missing fields and invalid enums) are handled in an SDK-specific way, with the bulk of the handling being done by generated code.
* Undetectable errors (field is present but types are mismatched) produce garbage data but otherwise will not crash or produce errors.

Ultimately, we'll have to do *something* if the user makes a non-compatible change to their schema, if possible I'd like to keep that behavior as consistent with other language SDKs as possible. Ideally I'd like to return an error for any error case, but if some errors are undetectable then keeping things consistent is the next best option.